src/KubernetesClient/Watch.cs

```diff
@@ -0,0 +1,491 @@
+            if (reader == null) CreateReader(await Response.Content.ReadAsStreamAsync().ConfigureAwait(false));
+            try
+            {
+                streamReader.CancelToken = cancelToken;
+#if !NET452
+                if (!await reader.ReadAsync(cancelToken).ConfigureAwait(false)) return null; // wait for the next event
+#else
+                if (!reader.Read()) return null; // wait for the next event
+#endif
+            }
+            catch (IOException ex) when (ex.InnerException is WebException wex && wex.Status == WebExceptionStatus.RequestCanceled)
+            {
+                throw new OperationCanceledException("The watch read was canceled.", ex.InnerException, cancelToken);
+            }
+
+            // deserialize the watch event
+            var e = new WatchEvent<T>();
+            bool gotType = false, gotObject = false;
+            while (true)
+            {
+#if !NET452
+                if (!await reader.ReadAsync(cancelToken).ConfigureAwait(false)) throw EOFError(); // move to the next property, if any
+#else
+                if (!reader.Read()) throw EOFError(); // move to the next property, if any
+#endif
+                if (reader.TokenType != JsonToken.PropertyName) break;
+                string name = (string)reader.Value;
+#if !NET452
+                if (!await reader.ReadAsync(cancelToken).ConfigureAwait(false)) throw EOFError(); // move to the property value
+#else
+                if (!reader.Read()) throw EOFError(); // move to the property value
+#endif
+                if (name == "type")
+                {
+                    e.Type = (WatchEventType)Enum.Parse(typeof(WatchEventType), (string)reader.Value, true);
+                    gotType = true;
+                }
+                else if (name == "object")
+                {
+                    if (!gotType) throw new JsonSerializationException("Expected type property before object.");
+                    if (e.Type != WatchEventType.Error) e.Object = serializer.Deserialize<T>(reader);
+                    else e.Error = serializer.Deserialize<V1Status>(reader);
+                    gotObject = true;
+                }
+                else
+                {
+#if !NET452
+                    await reader.SkipAsync(cancelToken).ConfigureAwait(false); // move to the next property, if any
+#else
+                    reader.Skip(); // move to the next property, if any
+#endif
+                }
+            }
+            if (!gotObject) throw new JsonSerializationException("The stream does not appear to contain watch events.");
+            return e;
+        }
+
+        #region CancelableStreamReader
+        // HACK: annoyingly, the TextReader interface doesn't support cancellation of async reads outside .NET core. furthermore, Json.NET
+        // doesn't support cancellation even in several versions of .NET core. so we have this hacky TextReader that attempts cancellation
+        sealed class CancelableStreamReader : TextReader
```

Comment:
what is new vs `PeekableStreamReader + CancelableStream` in WatcherDelegatingHandler

Reply:
It's not enough to have a cancelable stream. In fact, our underlying stream is already cancelable. What's needed is a cancelable TextReader, since the TextReader interface does not take cancellation tokens.

Reply:
The PeekableStreamReader doesn't support anything but peeking and reading lines, and doesn't support cancellation.